src/persist-client/src/internal/compact.rs

```diff
@@ -67,137 +69,161 @@ pub struct CompactRes<T> {
+                            }
+                        }
+                        Err(err) => {
+                            metrics.compaction.failed.inc();
+                            warn!("compaction for {} failed: {:#}", machine.shard_id(), err);
+                        }
+                    };
+                }
+                .instrument(compact_span)
                 .await;
 
-                metrics
-                    .compaction
-                    .seconds
-                    .inc_by(start.elapsed().as_secs_f64());
-
-                let res = match res {
-                    Ok(res) => res,
-                    Err(err) => {
-                        metrics.compaction.failed.inc();
-                        warn!("compaction for {} failed: {:#}", machine.shard_id(), err);
-                        return;
-                    }
-                };
-                let res = FueledMergeRes { output: res.output };
-                let applied = machine.merge_res(&res).await;
-                if applied {
-                    metrics.compaction.applied.inc();
-                } else {
-                    metrics.compaction.noop.inc();
-                    for part in res.output.parts {
-                        let key = part.key.complete(&machine.shard_id());
-                        retry_external(&metrics.retries.external.compaction_noop_delete, || {
-                            blob.delete(&key)
-                        })
-                        .await;
-                    }
-                }
+                // we can safely ignore errors here, it's possible the caller
+                // wasn't interested in waiting and dropped their receiver
+                let _ = completer.send(true);
             }
-            .instrument(compact_span),
-        ))
+        });
+
+        Compactor {
+            metrics,
+            sender: compact_req_sender,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Enqueues a [CompactReq] to be consumed by the compaction background task when available.
+    ///
+    /// Returns a receiver that indicates when compaction has completed. The receiver can be
+    /// safely dropped at any time if the caller does not wish to wait on completion.
+    pub fn compact_and_apply_background(
+        &self,
+        req: CompactReq<T>,
+    ) -> Option<oneshot::Receiver<bool>> {
+        let (compaction_completed_sender, compaction_completed_receiver) = oneshot::channel();
+        let new_compaction_sender = self.sender.clone();
+        let send = new_compaction_sender.send((req, compaction_completed_sender));
+
+        if let Err(e) = send {
+            // In the steady state we expect this to always succeed, but during
+            // shutdown it is possible the destination task has already spun down
+            warn!("compact_and_apply_background failed to send request: {}", e);
+            return None;
+        }
+
+        self.metrics.compaction.requested.inc();
```

Comment:
I think maybe we want to move this to the beginning of the method? that way if something happened to the channel sending, we'd see requested get incremented but not applied/noop/etc